### PR TITLE
Spot instances support for node pools

### PR DESCRIPTION
**Pull request by whites11**
This PR adds support for creating spot instances node pools.
Since I wasn't able to find any relevant field in the `AzureMachinePool` spec, I used 2 annotations in that CR for enabling node pools and specifying the max price desired.

The PR itself is quite easy, we can discuss tomorrow if we want to merge it or wait after the NP release.

### CODE REVIEW

File: service/controller/resource/nodepool/template/main.json [hunk 1]
```diff
@@ -284,6 +299,11 @@
         },
         "singlePlacementGroup": "[parameters('vmssSinglePlacementGroup')]",
         "virtualMachineProfile": {
+          "evictionPolicy": "Delete",
+          "priority": "[if(parameters('spotInstancesEnabled'), 'Spot', 'Regular')]",
+          "billingProfile": {
```
Comment (fiunchinho) [inline service/controller/resource/nodepool/template/main.json:304]: will this be ignored if `priority` is `regular`?

Comment (whites11) [reply service/controller/resource/nodepool/template/main.json:304]: Documentation says:

> Specifies the eviction policy for the Azure Spot virtual machine and Azure Spot scale set.

So I assume it is ignored if you don't have spot instances.
Plus in my tests I didn't notice issues.

**Review by fiunchinho — APPROVED** [review]
LGTM

File: service/controller/key/key.go [hunk 2]
```diff
@@ -620,6 +620,28 @@ func NodePoolDeploymentName(azureMachinePool *expcapzv1alpha3.AzureMachinePool)
...
+
+	return enabled
+}
+
+func NodePoolSpotInstancesMaxPrice(azuremachinePool *expcapzv1alpha3.AzureMachinePool) float64 {
+	maxPriceStr := azuremachinePool.Annotations[AnnotationSpotInstancesMaxPrice]
+	maxPrice, err := strconv.ParseFloat(maxPriceStr, 64)
+	if err != nil {
+		// Annotation not found or invalid.
+		return -1
```
Comment (tuommaki) [inline service/controller/key/key.go:639]: What happens when Azure is given `maxPrice: -1`?

Comment (whites11) [reply service/controller/key/key.go:639]: It automatically sets max price to the on demand price.
It's the default. See https://docs.microsoft.com/en-us/azure/templates/microsoft.compute/2019-12-01/virtualmachinescalesets#BillingProfile

**Review by tuommaki — APPROVED** [review]

**Comment by marians** [comment]
I just passed by this PR and was wondering about the terminology (using the term "instance" here in an Azure context. So far we have used the cloud provider specific terminology, even if it created a difference within our product, between providers. Have you had an internal discussion, has using "instances" been a conscious decision?

See also [this Slack thread](https://gigantic.slack.com/archives/C02521EC0/p1601544197065500) for alignment.

**Comment by whites11** [comment]
> I just passed by this PR and was wondering about the terminology (using the term "instance" here in an Azure context). So far we have used the cloud provider specific terminology, even if it created a difference within our product, between providers. Have you had an internal discussion, has using "instances" been a conscious decision?
> 
> See also [this Slack thread](https://gigantic.slack.com/archives/C02521EC0/p1601544197065500) for alignment.

Not that I paid any attention to the terminology here, but VMSS members are called instances in azure.

azure

**Comment by MartaD** [comment]
Feature was renamed from low-priority VMs to Spot VMs - let's use this term in docs.

**Pull request closed without merging**